### PR TITLE
[REG 2.067.0] Issue 14431 - huge slowdown of compilation speed

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=14431

In the pull request #4384, all instance has been changed to invoke
semantic3(). It was for the link-failure issue in specific case, but it
was too excessive.
1. Semantic analysis strategy for template instances:
   We cannot determine which instance does not need to be placed in object
   file until semantic analysis completed. Therefore, for all templates
   instantiated in root module, compiler should invoke their semantic3 --
   regardless of whether those are also instantiated in non-root module. If
   a template is _only_ instantiated in non-root module, we can elide its
   semantic3 (and for the compilation speed we should do that).
2. Code generation storategy for template instances:
   If a template is instantiated in non-root module, compiler usually does
   not have to put it in object file. But if a template is instantiated in
   both of root and non-root modules which mutually import each other, it
   needs to placed in objfile.

### CODE REVIEW

**Comment by WalterBright** [comment]
Does this replace https://github.com/D-Programming-Language/dmd/pull/4626 ?

**Comment by 9rnsr** [comment]
@WalterBright I'm not yet sure it's possible.

**Comment by 9rnsr** [comment]
OK, auto-tester is now all green.

**Comment by 9rnsr** [comment]
> Does this replace #4626 ?

Now, this actually replaces it.

**Comment by WalterBright** [comment]
> We cannot determine which instance does not need to be placed in object
 file until semantic analysis completed.

Why? If the semantic analysis produces different results, there's a bug somewhere.

> Therefore, for all templates instantiated in root module, compiler should invoke their semantic3 --
 regardless of whether those are also instantiated in non-root module.

I'm very unhappy with legitimizing the idea that this could produce different instantiations. The whole point is they should be the same.

> if a template is instantiated in both of root and non-root modules which mutually import each other, it
needs to placed in objfile.

This will simply kill compilation speed, as nobody seems to be able to resist having every module import every other module (Phobos is a prime example). I don't see how this PR fixes compile speed issues. I prefer the much more straightforward https://github.com/D-Programming-Language/dmd/pull/4780

**Comment by 9rnsr** [comment]
> Why? If the semantic analysis produces different results, there's a bug somewhere.

> I'm very unhappy with legitimizing the idea that this could produce different instantiations. The whole point is they should be the same.

In D2, we have attribute inference and it changes instantiated function mangled names. So, to get the exact lined symbol names, we need to run semantic3 on such the instances.

> This will simply kill compilation speed, as nobody seems to be able to resist having every module import every other module (Phobos is a prime example).

??? It's not a problem with the normal Phobos usage. For example, when you compile a "hello world" program, your code (root module) actually imports Phobos, but Phobos code (non-root module) does not import your program.

> I don't see how this PR fixes compile speed issues.

This PR stops excessive semantic3 for the templates instantiated in non-root module. It will restore the compilation speed up to the equivalent to the 2.065.

----

In my Windows 7 PC (64bit Core i7):
Compiled program: `import std.stdio; void main() { writeln("Hello world"); }`
Measurement by: `PS C:\d> Measure-Command {dmd.exe test.d}`

With 2.065: 800msec
With 2.066: 1.1sec
With 2.067: 1.4sec
With this PR: 800msec

**Comment by 9rnsr** [comment]
And, the excessive semantic3 had caused:
- redundant symbol emission (Issue 14508 - [REG2.067.0] compiling with -unittest instantiates templates in non-root modules)
- unnecessary -property check in Phobos code (Issue 14564 - [REG2.067] dmd -property -unittest combination causes compiler error)

This PR will also fix those regression issues.

**Comment by WalterBright** [comment]
> In D2, we have attribute inference and it changes instantiated function mangled names. So, to get the exact lined symbol names, we need to run semantic3 on such the instances.

There must never be a case where template function mangled names are the same but the instantiated bodies are different. This includes any attribute inference. If this happens, the whole compilation model will fall apart.

> For example, when you compile a "hello world" program, your code (root module) actually imports Phobos, but Phobos code (non-root module) does not import your program.

True, but every Phobos module imports every other Phobos module, and a user's project's modules will also typically import every other project module. Using separate compilation and heavy template use (which is becoming the norm) will become impractical.

> Measurement by:

Thank you for the good measurements. But I don't feel this is representative for projects outlined in the previous paragraph.

> redundant symbol emission (Issue 14508 - [REG2.067.0] compiling with -unittest instantiates templates in non-root modules)

Corrected as well with my simple PR https://github.com/D-Programming-Language/dmd/pull/4780

> unnecessary -property check in Phobos code (Issue 14564 - [REG2.067] dmd -property -unittest combination causes compiler error)

Hmm, that might be better addressed with -property fixes.

**Comment by JinShil** [comment]
> Hmm, that might be better addressed with -property fixes.

https://github.com/D-Programming-Language/dmd/pull/2305 would benefit from your feedback, and it can be rebased or modified per your input.

**Comment by 9rnsr** [comment]
> There must never be a case where template function mangled names are the same but the instantiated bodies are different. ...

You're misunderstanding my argument. If a template is instantiated in root module, we cannot skip its semantic3 until found the same instance in non-root. And *in worst case* the determination will be deferred until the global semantic3 stage completed, because a function defined in root module may import other modules from the function body.

(I noticed that, if the non-root instantiation found quickly, that would be able to stop being unneeded semantic3 sometimes. I think it would be a possible future optimization for the compilation speed.)

> True, but every Phobos module imports every other Phobos module, and a user's project's modules will also typically import every other project module. Using separate compilation and heavy template use (which is becoming the norm) will become impractical.

I noticed that #4780 would be disaster, Read [my comment](https://github.com/D-Programming-Language/dmd/pull/4780#issuecomment-117124587). 

> Corrected as well with my simple PR #4780

"simple solution" does not guarantee its correctness, especially for complex problems.

**Comment by 9rnsr** [comment]
Refactor code: Add `TemplateInstance::appendToModuleMember()`.

**Comment by MartinNowak** [comment]
Could someone draw a simple overview of all the cases to explain where the code of a template instance ends up.
It seems that we're adding more and more complexity, overfitting the code to every corner case instead of coming up with a general solution.
As the regression was already introduced in 2.066 we might take another release to solve this properly.

File: src/template.c [hunk 1]
```diff
@@ -5956,22 +5956,20 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
```
Comment (MartinNowak) [inline src/template.c]: What does that mean? If it has a name please abstract it.

File: src/template.c [hunk 2]
```diff
@@ -7302,6 +7244,72 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
...
+    if (mi)
+    {
+        if (mi->isRoot())
+        {
+            /* If both modules mi and md where the template is declared are
+             * roots, instead use md to store the generated code in it.
+             */
+            Module *md = tempdecl->getModule();
+            if (md && md->isRoot())
+                mi = md;
```
Comment (MartinNowak) [inline src/template.c]: That seems to be a new rule, why is it needed?
Who actually introduced that whacky minst concept, there can be multiple modules instantiating a template in D?

File: src/template.c (hunk 1 above)
Comment (9rnsr) [inline src/template.c]: The first instantiation `inst` was not instantiated in root module (`inst->minst == NULL` means it was speculative, and `inst->minst->isRoot() == false` means it was in non-root module), but `this` is instantiated in root module.

File: src/template.c (hunk 2 above)
Comment (9rnsr) [inline src/template.c]: > That seems to be a new rule, why is it needed?

By the `mi` adjustment, when a template `X` is declared in module `m.d`, and `m.d` is actually listed in command line, its instance will be constantly stored in `m.obj`.

> Who actually introduced that whacky minst concept, there can be multiple modules instantiating a template in D?

Walter had introduced the concept in #2550 (Later I renamed [instantiatingModule](https://github.com/D-Programming-Language/dmd/pull/2550/files#diff-46c6a9dc108e93f73a527f235351cfc4R68) to `minst`).
With a chain of template instantiations (A!() instantiates B!() instantiates C!() ...), when the first instance `A!()` exists in non-root module, all of their code generation can be elided. If `minst->isRoot() == false`, it means the instance is a part of non-root instantiation chain.

**Comment by 9rnsr** [comment]
OK, I succeeded fixing Win64 failure. All green.

**Comment by WalterBright** [comment]
> It seems that we're adding more and more complexity, overfitting the code to every corner case instead of coming up with a general solution.  As the regression was already introduced in 2.067.0 we might take another release to solve this properly.

This is why I introduced the alternate fix, https://github.com/D-Programming-Language/dmd/pull/4780

It is most appropriate for 2.068 at least, as it will enable us to get it out the door and has known behavior.

**Comment by 9rnsr** [comment]
> Could someone draw a simple overview of all the cases to explain where the code of a template instance ends up.

I wrote a wiki page for the explanation.
http://wiki.dlang.org/Template_Instantiation_Strategy

**Comment by WalterBright** [comment]
@9rnsr thank you for writing up the explanation. (It belongs in the source code, not the wiki.)

**Comment by WalterBright** [comment]
I still cannot say I understand this. I am not in favor of pulling this for 2.068 - I believe the potential for further regressions with the complex algorithm is high. Using https://github.com/D-Programming-Language/dmd/pull/4780 will revert to known behavior. I don't agree this is a 'disaster', the worst case is someone will have to throw the `-allinst` switch, and those cases don't come up all that often.

The big problem I'm having is D's templates are supposed to be simple to understand. These algorithms are very far from simple. I would like more time to try and completely understand what we are doing here - the complexity suggests to me that we have gone off the rails into a ditch somewhere. Things get more and more complex and yet the bug reports and regressions keep coming.

It's like the 'gagging' technique used in the compiler. It's an endless source of bugs and regressions. The whole design of that is a giant, steaming mass of corruption (and entirely my fault). I want to stop trying to fix each of the bugs in it, step back, and figure out a whole new approach to speculative compilation.

I believe that it would be most productive to defer such until after the dmd source is converted to D. I want that to be the first thing after the 2.068 release. The dmd source code in D will be much more tractable.

**Comment by 9rnsr** [comment]
Please show me any technical objections about this PR.

You are repeatedly talking about the regression risk. But how much the risk there is?
If you don't yet understand about the changes, I'm difficult to accept your risk talk.

I doubt that you are confusing simplicity of the concept and implementation. They're not same. A simple concept sometimes requires complex implementation - it's not rare in production software. (When those have equivalent usefulness, a complex implementation software with less bugs
is always preferred than a simple implementation software with many bugs.)

< rant > If you really want to keep implementation simple, please revert all my contributions around templates.
It will revert dmd code to the *known state*, and make the code simple by removing many corner case handlings, and finally reintroduce many link-failure issues. 
If we follow your words, it would be possible choice for the next release. Let's break everything! < /rant >

----

I think normally, I should have responsibility for all new bugs those introduced by my contributions.
My old bugfix had introduce regressions, so this PR fixes them.
If this PR introduces some regressions, I **will** fix them.

On the other hand, your "fix" will re-introduce link-failure bug and it's called a regression.
Honestly, can you really fix that issue with more better way?

I don't want to increase the number of bugs that won't be fixed.

**Comment by 9rnsr** [comment]
One more: Issue 14508 was a redundant symbol emission issue in object files. #4780 may fix it but also re-introduce link-failure problem. And #4780 has no effect to the compilation speed issue. Why the object file size issue is important than link failure bug? How do you think about issue 14431? Your priority is hard to understand to me.

**Comment by WalterBright** [comment]
> You are repeatedly talking about the regression risk. But how much the risk there is?

I believe it is substantial because this particular problem, and its various fixes, have repeatedly introduced regressions. It's gone on for years. This suggests to me that we do not completely understand the problem that needs solving. I would like to step back and think about it and come up with a way of fixing it for good.

> If this PR introduces some regressions, I will fix them.

I understand and appreciate that, and your history here proves that we can rely on you to do so.

> Let's break everything!

We all know you're our most valued contributor to DMD. We disagree on this one issue, not the rest of your contributions.

https://issues.dlang.org/show_bug.cgi?id=14431 should be corrected by https://github.com/D-Programming-Language/dmd/pull/4780, as it was caused by too many templates being compiled and emitted.

https://issues.dlang.org/show_bug.cgi?id=14508 The fixes there are mostly not affected by #4780.

> Your priority is hard to understand to me.

My priority is to get 2.068 out as a stable release. It's very late in the release cycle, it needs to go out.

Essentially, I believe we need to rethink when templates are instantiated. The current system is based on the C++ model (which doesn't have modules), with all kinds of patches laid on top of it. We can do better.

For example, you remarked earlier that semantic must be run on a template in order to determine its attributes and return type (for auto return), which is used to generated the mangled name, which is used to determine if a template has already been instantiated. That is correct. However, it is still true that it is a bug if semantic must be run in order to distinguish one template instantiation from another - the template arguments must be enough to uniquely determine the mangled name. Yet the mangled name includes the return type and the attributes. This suggests to me that the solution lies not in running semantic, but changing how the mangling is done.

**Comment by 9rnsr** [comment]
> I believe it is substantial because this particular problem, and its various fixes, have repeatedly introduced regressions. It's gone on for years.

Because we've continuously fixed remained bigger and more difficult issues, than the issues we had already fixed number of years ago, to increase compiler stability.
For example: #3384 was a recent big fix for the remaining forward reference issue around class inheritance. It had introduced some regressions, but finally we've actually fixed one of the remained big issues.

> This suggests to me that we do not completely understand the problem that needs solving. I would like to step back and think about it and come up with a way of fixing it for good.

Yes we're not perfect. However our life are finite, so we should also consider the cost-effectiveness of our time.

> I understand and appreciate that, and your history here proves that we can rely on you to do so.

Thank you very much.

> https://issues.dlang.org/show_bug.cgi?id=14508 The fixes there are mostly not affected by #4780.

I cannot understand this. Are you saying the summary of #4780 is incorrect? It's titled with "Issue 14508"...

> My priority is to get 2.068 out as a stable release. It's very late in the release cycle, it needs to go out.

At least this fix is a part of the stabilization. This PR does not suggest any completely new feature. This PR is based on the current template instantiation model (it's introduced by #2550 so since 2.064).

> Essentially, I believe we need to rethink when templates are instantiated. ...

Rethinking about the new mangling rule is good. But before than that, we should consider to fix **existing** bugs coming from the current mangling rule - for example [issue 10619](https://issues.dlang.org/show_bug.cgi?id=10619) and [issue 14739](https://issues.dlang.org/show_bug.cgi?id=14739)).

**Comment by MartinNowak** [comment]
> I wrote a wiki page for the explanation.
> http://wiki.dlang.org/Template_Instantiation_Strategy

Thanks for the nice summary.

It seems that by solving the mutual import issue we'd at least have a stable rule that allows efficient compilation without linker errors.
Granted, it's a complex solution, but it works for the time being. The root vs. non-root distinction introduced by #2550 comes with a lot of intrinsic complexity and has caused much grief, but also came with some good compile time/object size improvements. We shouldn't throw this away, now it's almost finished.

The simple and failsafe solution is to emit code to every object that uses an instance, letting the linker sort out the duplicates. If we can get this fast enough it might be an alternative, particularly for the multiple module to multiple object file compilation mode.

File: src/template.c (hunk 1 above)
Comment (MartinNowak) [inline src/template.c]: A simple `bool isSpeculative() { return minst == NULL; }` would help.

Comment (MartinNowak) [inline src/template.c]: So this selects the case where `this` is non-speculative and root, but the same instance `inst` was done in speculative or non-root context. Why would you want non-root instances here?

File: src/template.c [hunk 3]
```diff
@@ -5958,22 +5958,23 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
...
             TemplateInstance *ti = tinst;
-            while (ti && ti != inst)
-                ti = ti->tinst;
-            if (ti != inst)    // Bugzilla 13379: Prevent circular chain
-                inst->tinst = tinst;
-        }
+            minst = inst->minst;
+            tinst = inst->tinst;
+            inst->minst = mi;
+            inst->tinst = ti;
```
Comment (MartinNowak) [inline src/template.c]: Why swapping the contexts and appending `inst` instead of simply continuing with `this` which should already be member of a root module?

**Comment by MartinNowak** [comment]
> In D2, we have attribute inference and it changes instantiated function mangled names. So, to get the exact lined symbol names, we need to run semantic3 on such the instances.

Since when does inference affect mangling? I thought we'd avoided that b/c of ABI stability and separate compilation concerns.

File: src/template.c (hunk 3 above)
Comment (MartinNowak) [inline src/template.c]: It's a bad idea for debugging to change minst/tinst, just like you wouldn't want to change the scope of a symbol.

File: src/template.c (hunk 1 above)
Comment (9rnsr) [inline src/template.c]: If `inst` was done in speculative, it's already inserted in root module members. So the second time insertion is necessary only for non-root instances.

File: src/template.c (hunk 3 above)
Comment (9rnsr) [inline src/template.c]: We need the second insertion at most only once. By the swap, `inst` is changed to a root instance, then the condition at line 5961 no longer hit. This is for the calculation amount reduction.

And, `minst` and `tinst` do not affect semantic analysis results. So the swap is not dangerous.

**Comment by 9rnsr** [comment]
> Since when does inference affect mangling? I thought we'd avoided that b/c of ABI stability and separate compilation concerns.

Since the first (inference modifies `FuncDeclaration::type`, then it will be a part of function symbol). In my recognition, typesafe linking is priority than the compilation speed currently.

File: src/template.c (hunk 1 above)
Comment (9rnsr) [inline src/template.c]: > A simple bool isSpeculative() { return minst == NULL; } would help.

Unfortunately we already have `Dsymbol::isSpeculative()` and its meaning is not same with the condition tested in here. I'd really want to determine good name, but it's not discovered yet.

File: src/template.c (hunk 3 above)
Comment (MartinNowak) [inline src/template.c]: I still don't get it.
The whole template compilation/codegen pipeline is already complicated enough. We should avoid additional tricks, even more so if they are undocumented.

File: src/template.c [hunk 4]
```diff
@@ -5958,22 +5958,23 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
         {
-            // Reconnect the chain if this instantiation is not in speculative context.
+            // Swap the minst and tinst between inst and this.
```
Comment (MartinNowak) [inline src/template.c]: You should comment _why_ you're doing something. _What_ you're doing is fairly obvious.

File: src/template.c (hunk 1 above)
Comment (MartinNowak) [inline src/template.c]: > If inst was done in speculative, it's already inserted in root module members.

Why is that???

File: src/template.c (hunk 3 above)
Comment (9rnsr) [inline src/template.c]: I updated to make the comment detailed more. Is that enough?

A poor solution is adding one bool flag `bool TemplateInstance::isAlreadyInstantiatedInRoot;` and test&check it here. But it will increment the size of TemplateInstance object. As far as possible I'd like to save memory footprint of dmd.

File: src/template.c (hunk 1 above)
Comment (9rnsr) [inline src/template.c]: If a template is instantiated in speculative context, the context should handle *all* errors during the instantiation - it's including the errors in function body.

```d
void foo(T)(T t) { auto x = undefined_name; }
static assert(!is(typeof(foo(1))));   // the undefined error should be captured in here.
```

For that, all of speculative instantiations will be handled as root instances, and then, their semantic3 pass will surely get called.

File: src/template.c (hunk 3 above)
Comment (MartinNowak) [inline src/template.c]: > As far as possible I'd like to save memory footprint of dmd.

You can get that for free by combining all the boolean flags to a FLAGS enum.
There is currently a hole for one bool anyhow ;).
https://github.com/D-Programming-Language/dmd/blob/61f8146e94dca8d3a0617d7235359b8e8b53e4b2/src/template.h#L312

Comment (9rnsr) [inline src/template.c]: I know the holes between bool variables, but it's unrelated refactoring, and should not go for the release.

Why you require such the redundant bool? Current "swap" is enough smart and efficient. I have no reason to degrade code quality.

**Comment by 9rnsr** [comment]
@MartinNowak @WalterBright Do you have any other objections about this PR? Even if you don't want to go this change in 2.068 release, we can fix the issue in master branch...

**Comment by WalterBright** [comment]
The algorithm is very difficult to understand. I don't know how to explain it to someone else. I worry that if the only person it is understandable to is Kenji, then nobody else can maintain the compiler. This is not practical for a team organization.

1. Martin, do you understand the algorithm?
2. The wiki, which refers to compiler internals, does not belong on the wiki, it belongs in the compiler source code.
3. The wiki should have an explanation for how it works that is intended for the D user, i.e. not in terms of compiler internals.

**Comment by 9rnsr** [comment]
@WalterBright I placed the text in wiki because it was convenient. I'll move it in dmd repo if you hope that (e.g. `dmd/doc/note.txt`?).

**Comment by 9rnsr** [comment]
> I worry that if the only person it is understandable to is Kenji, then nobody else can maintain the compiler.

I recall that David Nadlinger has fixed a regression [issue 13478](https://issues.dlang.org/show_bug.cgi?id=13478) in 2.066.1 devel, which was the bug in the template instantiation code.

@klickverbot Could you please join the review of this PR?

File: src/template.c (hunk 1 above)
Comment (MartinNowak) [inline src/template.c]: That's actually very simple and roughly means `if (this->needsCodegen && !inst->needsCodegen)`.

File: src/template.c (hunk 3 above)
Comment (MartinNowak) [inline src/template.c]: This is a really good argument to split TemplateInstantiation and TemplateInstance. This double usage had always confused me and likely eats up a lot of memory as well.

Comment (MartinNowak) [inline src/template.c]: So the point of this is that all identical TemplateInstances delegate to the first TemplateInstance (findExistingInstance) to avoid duplicate semantic. But if the first (primary) instance was done in speculative context, then no code will be generated. Therefore `tinst` and `minst` are swapped to make the primary instance a normal root instance.
It's still confusing that the scopes of `this` and `inst` won't match their `tinst` and `minst` variables.
It would be nice to move this into a shallowSwap function and to swap all fields that deal with the instantiation but not the instance, e.g. swap scope but not members.

Comment (MartinNowak) [inline src/template.c]: Will the fact that a parent was a speculative template instance affect the semantic of members?
It should only change the error reporting, but maybe I'm overlooking something here.

File: src/template.c [hunk 5]
```diff
@@ -7335,6 +7292,83 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
...
+    Module *mi = minst;
+    if (mi)
+    {
+        if (mi->isRoot())
+        {
+            /* If both modules mi and md where the template is declared are
+             * roots, instead use md to store the generated code in it.
+             */
+            if (md && md->isRoot())
+                mi = md;
```
Comment (MartinNowak) [inline src/template.c]: Again why this change? It seems unrelated to the fix.

File: src/template.c [hunk 6]
```diff
@@ -7335,6 +7292,83 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
...
+    for (size_t i = 0; 1; i++)
+    {
+        if (i == a->dim)
+        {
+            a->push(this);
+            if (mi->semanticRun >= PASSsemantic3done && mi->isRoot())
+                Module::addDeferredSemantic3(this);
+            break;
+        }
+        if (this == (*a)[i])    // if already in Array
```
Comment (MartinNowak) [inline src/template.c]: Why would it already be in the array?
Seems like noone should call appendToModuleMember multiple times or for TIs that already are root module members.

File: src/struct.c [hunk 7]
```diff
@@ -100,8 +100,12 @@ void semanticTypeInfo(Scope *sc, Type *t)
 
             // If the struct is in a non-root module, run semantic3 to get
             // correct symbols for the member function.
-            // Note that, all instantiated symbols will run semantic3.
-            if (sd->inNonRoot())
+            if (TemplateInstance *ti = sd->isInstantiated())
+            {
+                if (!ti->minst || !ti->minst->isRoot())
```
Comment (MartinNowak) [inline src/struct.c]: Why is this necessary for speculative instances?

File: src/template.c (hunk 1 above)
Comment (9rnsr) [inline src/template.c]: Yes, that's right.

**Comment by MartinNowak** [comment]
> In D2, we have attribute inference and it changes instantiated function mangled names. So, to get the exact lined symbol names, we need to run semantic3 on such the instances.

I still don't understand how semantic3 for correct mangling is necessary to find existing instances.

File: src/template.c (hunk 3 above)
Comment (9rnsr) [inline src/template.c]: > It's still confusing that the scopes of this and inst won't match their tinst and minst variables.

The scope of instantiated position, it's the `Scope*sc` parameter of `TemplateInstance::semantic()` function, is not saved anywhere. Therefore the mismatch of tinst/minst won't leak out of the function. And in `TemplateInstance::semantic2()` and `semantic3()`, the given `Scope *sc` parameter is just ignored.

> Will the fact that a parent was a speculative template instance affect the semantic of members?

Never.

File: src/template.c (hunk 5 above)
Comment (9rnsr) [inline src/template.c]: This is not necessary to fix the regression, but it will increase stability for the position where the instance will be inserted. For example:

a.d
```d
import ...;

template X() { ... }

X!()
```

b.d
```d
import a, ...;

X!()
```

With the command line `dmd -c a.d b.d`:

- If this adjustment does not exist, the instance X!() will be inserted to the member of *either* `a` or `b`. The actual position will depend on the semantic order.
- If this adjustment exists, the instance X!() will be *always* inserted to the member of `a`.

By that, the `X!()` code will be always stored into `a.obj`, in that case. I think the stable algorithm result is better than unstable, if there's not so much performance difference.

File: src/template.c (hunk 6 above)
Comment (9rnsr) [inline src/template.c]: I also think the conflict would not happen, but I'm really not sure it yet, because same check existed in old code. So I'd just keep it until we can sure it is unnecessary.

**Comment by 9rnsr** [comment]
> I still don't understand how semantic3 for correct mangling is necessary to find existing instances.

Perhaps we can elide semantic3 call from some root instances. But currently semantic3 is always called on all root instances. We cannot improve it yet.

File: src/struct.c (hunk 7 above)
Comment (9rnsr) [inline src/struct.c]: I don't recall the reason...
Sure, when `ti` is speculative, doing `Module::addDeferredSemantic3(sd)` would be redundant, because it's already done in `TemplateInstance::appendToModuleMember`.
I'll update it.

Comment (9rnsr) [inline src/struct.c]: Done.

**Comment by WalterBright** [comment]
> I still don't understand how semantic3 for correct mangling is necessary to find existing instances.

It isn't necessary. A template's arguments completely determine it. If the arguments match the arguments for existing instances, then it MUST MATCH and running semantic3 is not necessary.

**Comment by 9rnsr** [comment]
I added test case to show that the issue 14901 will be actually fixed.

**Comment by MartinNowak** [comment]
I know it's a complex change but all of the fixes here make sense and solve a few big regressions.
Let's please go on with this and work on cleaning up the template implementation later.

**Comment by MartinNowak** [comment]
> It isn't necessary. A template's arguments completely determine it. If the arguments match the arguments for existing instances, then it MUST MATCH and running semantic3 is not necessary.

I agree, but changing this isn't actually part of this PR.

**Comment by MartinNowak** [comment]
Can you please rebase+reopen this against stable. Please do so soon, so that we can rigorously test this before 2.068.1.

**Comment by 9rnsr** [comment]
Reopened against stable branch. See #4944.
